osu.Game/Rulesets/Edit/Checks/CheckHitsoundsFormat.cs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using ManagedBass;
+using osu.Framework.Audio.Callbacks;
+using osu.Game.Beatmaps;
+using osu.Game.Extensions;
+using osu.Game.Rulesets.Edit.Checks.Components;
+
+namespace osu.Game.Rulesets.Edit.Checks
+{
+    public class CheckHitsoundsFormat : ICheck
+    {
+        public CheckMetadata Metadata => new CheckMetadata(CheckCategory.Audio, "Checks for hitsound formats.");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateFormatUnsupported(this),
+            new IssueTemplateIncorrectFormat(this),
+        };
+
+        private IEnumerable<ChannelType> allowedFormats => new ChannelType[]
+        {
+            ChannelType.WavePCM,
+            ChannelType.WaveFloat,
+            ChannelType.OGG,
+            ChannelType.Wave | ChannelType.OGG,
+        };
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            var beatmapSet = context.Beatmap.BeatmapInfo.BeatmapSet;
+            var audioFile = beatmapSet?.GetFile(context.Beatmap.Metadata.AudioFile);
+
+            if (beatmapSet == null) yield break;
+
+            foreach (var file in beatmapSet.Files)
+            {
+                if (audioFile != null && file.File == audioFile.File) continue;
+
+                using (Stream data = context.WorkingBeatmap.GetStream(file.File.GetStoragePath()))
+                {
+                    if (data == null)
+                        continue;
+
+                    var fileCallbacks = new FileCallbacks(new DataStreamFileProcedures(data));
+                    int decodeStream = Bass.CreateStream(StreamSystem.NoBuffer, BassFlags.Decode, fileCallbacks.Callbacks, fileCallbacks.Handle);
+
+                    // If the format is not supported by BASS
+                    if (decodeStream == 0)
+                    {
+                        if (AudioCheckUtils.HasAudioExtension(file.Filename) && probablyHasAudioData(data))
+                            yield return new IssueTemplateFormatUnsupported(this).Create(file.Filename);
+
+                        continue;
+                    }
+
+                    var audioInfo = Bass.ChannelGetInfo(decodeStream);
+
+                    if (!allowedFormats.Contains(audioInfo.ChannelType))
+                        yield return new IssueTemplateIncorrectFormat(this).Create(file.Filename, audioInfo.ChannelType.ToString());
```

Comment:
Is the channel type going to even be user-printable here? It's just doing `ToString()` on an enum which means it'll likely turn up weird gibberish?

Reply:
I would say so yes.